### DISCUSSION TITLE
Proposal: Multiple personalities

### DISCUSSION THREAD

**Discussion by robstumborg** [Robitx/gp.nvim#40] [Ideas]
What do you think about having pre-configured "personalities" that could be used with the plugin?

Sometimes I think it would be nice to use this plugin/neovim for general OpenAI use and not necessarily just for assistance with programming, but switching the config every time would be too cumbersome.

Here's an example of what a configuration could look like:

```lua
local conf = {
	chat_personalities = {

		programmer = {
			chat_model = { model = "gpt-3.5-turbo-16k", temperature = 1.1, top_p = 1 },
			chat_system_prompt = "You are a general AI assistant for a computer programmer.",
			chat_custom_instructions = "The user provided the additional info about how they would like you to respond:\n\n"
				.. "- If you're unsure don't guess and say you don't know instead.\n"
				.. "- Ask question if you need clarification to provide better answer.\n"
				.. "- Think deeply and carefully from first principles step by step.\n"
				.. "- Zoom out first to see the big picture and then zoom in to details.\n"
				.. "- Use Socratic method to improve your thinking and coding skills.\n"
				.. "- Don't elide any code from your output if the answer requires coding.\n"
				.. "- Take a deep breath; You've got this!\n",
		},

		writer = {
			chat_model = { model = "gpt-3.5-turbo-16k", temperature = 1.1, top_p = 1 },
			chat_system_prompt = "You are a general AI assistant for a writer.",
			chat_custom_instructions = "Some specific instructions that are useful for somebody writing a novel:\n\n"
        .. " - blah\n"
        .. " - blah\n"
        .. " - blah\n"
		},

    technical_writer = {
      chat_model = { model = "gpt-3.5-turbo-16k", temperature = 1.1, top_p = 1 },
      chat_system_prompt = "You are a general AI assistant for somebody writing technical documentation for computer software.",
      chat_custom_instructions = "Some specific instructions that are useful for somebody writing technical docs:\n\n"
        .. " - blah\n"
        .. " - blah\n"
        .. " - blah\n"
    }
	},

  default_chat_personality = "programmer",
}
```

Switching between personalities could be a command like `:GpChatPersonality writer`.

I'd be willing to put in the time to create this functionality if you think it's something you want in the plugin.

**Comment by Robitx** [comment]
@robstumborg Will do. :slightly_smiling_face:

**Reply by Robitx** [reply]
@robstumborg sorry it took a month. I was partially avoiding it, because of associated config deprecations.

**Reply by robstumborg** [reply]
@Robitx that's amazing you did it. no apology necessary. thank you! I am going to be using this starting now.

**Comment by qaptoR** [comment]
how do we access this feature? I don't see it documented anywhere in my recent usage.

**Reply by qaptoR** [reply]
oh wait.. I think I see how you implemented this.

![image](https://github.com/user-attachments/assets/a5484556-7a02-43b5-b460-05d701ad1202)

So, is the 'system prompt' part of the agent the way that personalities are currently supported?

If so, I'd like to update this proposal to suggest that we have 3 different config sections
- `providers {...}`
- `agents {...}`
- `roles {...}`

I would prefer this because I currently have 13 different agents (2 model sizes x 5 providers + 3 more for perplexity online) x code and chat versions.
My list of agents would grow way too fast if for every model I wanted to use I had to create every version of it's personality.
We should be able to mix/match models and roles on the fly, with a picker for roles similar to the one for models.

additionally, temperature and top_p should be configurable outside of the config, separate from the model. I think it would make more sense to do the following (or something as equally or more controllable):
- have a default chat and code top_p/temp applied to all models
- roles have an optional top_p/temp that overrides the default
- there is a toggle on/off `override top_p` and `override temp` 
  - there is also a `GpSetTop_pOverride` command
